File: packages/claude-agent-sdk-adapter-python/src/agentmark_claude_agent_sdk/tool_registry.py
"""Tool registry for Claude Agent SDK adapter.

Ported from TypeScript: packages/claude-agent-sdk-adapter/src/tool-registry.ts
"""

from __future__ import annotations

import asyncio
import inspect
from dataclasses import dataclass
from typing import Any

from .types import AsyncToolFunction, ToolFunction


@dataclass
class RegisteredToolExecutor:
    """A registered tool executor wrapper."""

    name: str
    """Tool name."""

    _fn: ToolFunction | AsyncToolFunction
    """The registered function."""

    async def execute(self, args: dict[str, Any], ctx: dict[str, Any] | None = None) -> Any:
        """Execute the tool function.

        Args:
            args: Tool arguments.
            ctx: Optional context.

        Returns:
            Tool execution result.
        """
        result = self._fn(args, ctx)
        if inspect.iscoroutine(result) or asyncio.isfuture(result):
            return await result
        return result


class ClaudeAgentToolRegistry:
    """Type-safe tool registry for Claude Agent SDK adapter.

    Tools registered here are bridged to the Claude Agent SDK via MCP.
    Tool descriptions and parameters are derived from the prompt's tool schema.

    Example:
        tool_registry = ClaudeAgentToolRegistry()
        tool_registry.register("search_documents", async_search_fn)
        tool_registry.register("get_weather", get_weather_fn)
    """

    def __init__(self) -> None:
        """Create a new tool registry."""
        self._map: dict[str, ToolFunction | AsyncToolFunction] = {}

    def register(
        self,
        name: str,
        fn: ToolFunction | AsyncToolFunction,
    ) -> ClaudeAgentToolRegistry:
        """Register a tool execution function.

        Tool description and parameters are derived from the prompt's tool schema.

        Args:
            name: Unique tool name (must match tool name in prompt config).
            fn: Function to execute the tool (sync or async).

        Returns:
            self for chaining.
        """
        self._map[name] = fn
        return self

    def get(self, name: str) -> RegisteredToolExecutor | None:
        """Get a tool executor by name.

        Args:
            name: Tool name to retrieve.

        Returns:
            RegisteredToolExecutor or None if not found.
        """
        fn = self._map.get(name)
        if fn is None:
            return None
        return RegisteredToolExecutor(name=name, _fn=fn)

    def has(self, name: str) -> bool:
        """Check if a tool is registered.

        Args:
            name: Tool name to check.

        Returns:
            True if the tool is registered.
        """
        return name in self._map

    @property
    def size(self) -> int:
        """Get the number of registered tools."""
        return len(self._map)

    def list_names(self) -> list[str]:
        """Get all tool names.

        Returns:
            List of registered tool names.
        """
        return list(self._map.keys())

File: packages/claude-agent-sdk-adapter/src/tool-registry.ts
/**
 * Merge type for accumulating tool types during chaining.
 */
type Merge<A, B> = {
  [K in keyof A | keyof B]: K extends keyof B
    ? B[K]
    : K extends keyof A
      ? A[K]
      : never;
};

/**
 * Type-safe tool registry for Claude Agent SDK adapter.
 *
 * Tools registered here are bridged to the Claude Agent SDK via MCP.
 * Tool descriptions and parameters are derived from the prompt's tool schema.
 *
 * @example
 * ```typescript
 * const toolRegistry = new ClaudeAgentToolRegistry()
 *   .register("search_documents", async ({ query }) => {
 *     const results = await searchDB(query);
 *     return { results };
 *   })
 *   .register("get_weather", async ({ location }) => {
 *     return await fetchWeather(location);
 *   });
 * ```
 */
export class ClaudeAgentToolRegistry<
  TD extends { [K in keyof TD]: { args: any } },
  RM extends Partial<Record<keyof TD, any>> = {}
> {
  /**
   * Phantom type for tracking tool types.
   * This enables TypeScript to infer correct types for tool inputs/outputs.
   */
  declare readonly __tools: { input: TD; output: RM };

  private map: {
    [K in keyof TD]?: (args: TD[K]["args"]) => any;
  } = {};

  /**
   * Register a tool execution function.
   * Tool description and parameters are derived from the prompt's tool schema.
   *
   * @param name - Unique tool name (must match tool name in prompt config)
   * @param fn - Async function to execute the tool
   * @returns New registry with accumulated tool type
   */
  register<K extends keyof TD, R>(
    name: K,
    fn: (args: TD[K]["args"]) => R
  ): ClaudeAgentToolRegistry<TD, Merge<RM, { [P in K]: R }>> {
    this.map[name] = fn;
    return this as unknown as ClaudeAgentToolRegistry<
      TD,
      Merge<RM, { [P in K]: R }>
    >;
  }

  /**
   * Get a tool execution function by name.
   */
  get<K extends keyof TD & keyof RM>(name: K) {
    return this.map[name] as (args: TD[K]["args"]) => RM[K];
  }

  /**
   * Check if a tool is registered.
   */
  has(name: string): boolean {
    return name in this.map;
  }

  /**
   * Get the number of registered tools.
   */
  get size(): number {
    return Object.keys(this.map).length;
  }

  /**
   * Get all tool names.
   */
  getToolNames(): string[] {
    return Object.keys(this.map);
  }
}
